src/Stratis.Bitcoin.Features.Consensus/Rules/CommonRules/PowCoinviewRule.cs

```diff
@@ -1,102 +1,72 @@
+        {
+            this.Logger.LogTrace("()");
 
-        /// <summary>Provider of time functions.</summary>
-        protected readonly IDateTimeProvider dateTimeProvider;
+            this.consensusParams = this.Parent.Network.Consensus;
+            this.consensusOptions = this.consensusParams.Option<PowConsensusOptions>();
 
-        /// <summary>
-        /// Initializes a new instance of the <see cref="PowConsensusValidator"/> class.
-        /// </summary>
-        /// <param name="network">Specification of the network the node runs on - regtest/testnet/mainnet.</param>
-        /// <param name="checkpoints">Provider of block header hash checkpoints.</param>
-        /// <param name="dateTimeProvider">Provider of time functions.</param>
-        /// <param name="loggerFactory">Factory for creating loggers.</param>
-        public PowConsensusValidator(Network network, ICheckpoints checkpoints, IDateTimeProvider dateTimeProvider, ILoggerFactory loggerFactory)
-        {
-            Guard.NotNull(network, nameof(network));
-            Guard.NotNull(network.Consensus.Option<PowConsensusOptions>(), nameof(network.Consensus.Options));
-
-            this.logger = loggerFactory.CreateLogger(this.GetType().FullName);
-            this.ConsensusParams = network.Consensus;
-            this.ConsensusOptions = network.Consensus.Option<PowConsensusOptions>();
-            this.network = network;
-            this.dateTimeProvider = dateTimeProvider;
-            this.PerformanceCounter = new ConsensusPerformanceCounter(this.dateTimeProvider);
-            this.Checkpoints = checkpoints;
+            this.Logger.LogTrace("(-)");
         }
 
         /// <inheritdoc />
-        public virtual void ExecuteBlock(RuleContext context, TaskScheduler taskScheduler = null)
+        public override async Task RunAsync(RuleContext context)
         {
-            this.logger.LogTrace("()");
+            this.Logger.LogTrace("()");
 
             Block block = context.BlockValidationContext.Block;
             ChainedHeader index = context.BlockValidationContext.ChainedHeader;
             DeploymentFlags flags = context.Flags;
             UnspentOutputSet view = context.Set;
 
-            this.PerformanceCounter.AddProcessedBlocks(1);
-            taskScheduler = taskScheduler ?? TaskScheduler.Default;
+            this.Parent.PerformanceCounter.AddProcessedBlocks(1);
 
             long sigOpsCost = 0;
             Money fees = Money.Zero;
             var checkInputs = new List<Task<bool>>();
             for (int txIndex = 0; txIndex < block.Transactions.Count; txIndex++)
             {
-                this.PerformanceCounter.AddProcessedTransactions(1);
+                this.Parent.PerformanceCounter.AddProcessedTransactions(1);
                 Transaction tx = block.Transactions[txIndex];
                 if (!context.SkipValidation)
                 {
-                    
+                    // TODO: Simplify this condition.
```

Comment:
See issue https://github.com/stratisproject/StratisBitcoinFullNode/issues/1433  for the improvement suggestion